Mathlib/Geometry/Euclidean/Sphere/Tangent.lean

```diff
@@ -0,0 +1,405 @@
+  {as ∈ s.tangentSet | p ∈ as}
+
+lemma mem_tangentsFrom_iff {as : AffineSubspace ℝ P} {s : Sphere P} {p : P} :
+    as ∈ s.tangentsFrom p ↔ as ∈ s.tangentSet ∧ p ∈ as :=
+  Iff.rfl
+
+lemma mem_tangentSet_of_mem_tangentsFrom {as : AffineSubspace ℝ P} {s : Sphere P} {p : P}
+    (h : as ∈ s.tangentsFrom p) : as ∈ s.tangentSet :=
+  h.1
+
+lemma mem_of_mem_tangentsFrom {as : AffineSubspace ℝ P} {s : Sphere P} {p : P}
+    (h : as ∈ s.tangentsFrom p) : p ∈ as :=
+  h.2
+
+lemma isTangent_of_mem_tangentsFrom {as : AffineSubspace ℝ P} {s : Sphere P} {p : P}
+    (h : as ∈ s.tangentsFrom p) : s.IsTangent as :=
+  isTangent_of_mem_tangentSet h.1
+
+/-- The set of all maximal common tangent spaces to the spheres `s₁` and `s₂`. -/
+def commonTangents (s₁ s₂ : Sphere P) : Set (AffineSubspace ℝ P) :=
+  s₁.tangentSet ∩ s₂.tangentSet
+
+lemma mem_commonTangents_iff {as : AffineSubspace ℝ P} {s₁ s₂ : Sphere P} :
+    as ∈ s₁.commonTangents s₂ ↔ as ∈ s₁.tangentSet ∧ as ∈ s₂.tangentSet :=
+  Iff.rfl
+
+lemma commonTangents_comm (s₁ s₂ : Sphere P) : s₁.commonTangents s₂ = s₂.commonTangents s₁ :=
+  Set.inter_comm _ _
+
+/-- The set of all maximal common internal tangent spaces to the spheres `s₁` and `s₂`: tangent
+spaces containing a point weakly between the centers of the spheres. -/
+def commonIntTangents (s₁ s₂ : Sphere P) : Set (AffineSubspace ℝ P) :=
+  {as ∈ s₁.commonTangents s₂ | ∃ p ∈ as, Wbtw ℝ s₁.center p s₂.center}
+
+/-- The set of all maximal common external tangent spaces to the spheres `s₁` and `s₂`: tangent
+spaces not containing a point strictly between the centers of the spheres. (In the degenerate case
+where the two spheres are the same sphere with radius 0, the space is considered both an internal
+and an external common tangent.) -/
+def commonExtTangents (s₁ s₂ : Sphere P) : Set (AffineSubspace ℝ P) :=
+  {as ∈ s₁.commonTangents s₂ | ∀ p ∈ as, ¬Sbtw ℝ s₁.center p s₂.center}
+
+lemma mem_commonIntTangents_iff {as : AffineSubspace ℝ P} {s₁ s₂ : Sphere P} :
+    as ∈ s₁.commonIntTangents s₂ ↔
+      as ∈ s₁.commonTangents s₂ ∧ ∃ p ∈ as, Wbtw ℝ s₁.center p s₂.center :=
+  Iff.rfl
+
+lemma mem_commonExtTangents_iff {as : AffineSubspace ℝ P} {s₁ s₂ : Sphere P} :
+    as ∈ s₁.commonExtTangents s₂ ↔
+      as ∈ s₁.commonTangents s₂ ∧ ∀ p ∈ as, ¬Sbtw ℝ s₁.center p s₂.center :=
+  Iff.rfl
+
+@[simp] lemma union_commonIntTangents_commonExtTangents (s₁ s₂ : Sphere P) :
+    s₁.commonIntTangents s₂ ∪ s₁.commonExtTangents s₂ = s₁.commonTangents s₂ := by
+  ext as
+  rw [Set.mem_union, mem_commonIntTangents_iff, mem_commonExtTangents_iff, ← and_or_left,
+    and_iff_left_iff_imp]
+  rintro -
+  by_cases h : ∃ p ∈ as, Wbtw ℝ s₁.center p s₂.center
+  · exact .inl h
+  · refine .inr ?_
+    simp_rw [not_exists, not_and] at h
+    rintro p hp
+    exact mt Sbtw.wbtw (h p hp)
+
+/-- The spheres `s₁` and `s₂` are externally tangent at the point `p`. -/
+structure IsExtTangentAt (s₁ s₂ : Sphere P) (p : P) : Prop where
```

Comment:
Could you please add a lemma about how this definition (and the 3 below) interact with the sets above?

If I understand correctly, two spheres are externally tangent if `s₁.commonIntTangents s₂` is nonempty, and they are internally tangent if `s₁.commonExtTangents s₂` is nonempty.

Reply:
Ignoring degenerate cases, two spheres have common external tangents unless one is strictly inside the other. They have common internal tangents if they lie outside each other except possibly for a single point of tangency. If they are externally tangent, they have a single common internal tangent. If they are internally tangent (and not the same sphere), they have a single common external tangent. Otherwise, sphere with common internal / external tangents have two in the plane (reflect in the line of centers) or infinitely many in higher dimensions.

The definitions are expected to be used for problems such as IMO 2008 P6, which ends "Prove that the common external tangents of ω_1 and ω_2 intersect on ω."

Reply:
Got it. Do you think formal proofs of those statements are useful API lemmas? Or would you rather postpone adding them until they are needed?